### ISSUE TITLE
Failed to infer types during nested generic method invocation

### ISSUE THREAD

**Issue by testforstephen** [eclipse-jdt/eclipse.jdt.core#506]
```java
import java.util.function.Function;

class Convert<A> {
    public static <B> void test(final B a) {
        Convert<B> res1 = convert(arg -> create(arg), a); // ok

        Convert<B> res2 = convert(arg -> wrap(create(arg)), a); // error: Type mismatch
    }

    public static <C, D> Convert<D> convert(final Function<C, Convert<D>> func, final C value) {
        return null;
    }

    public static <E> E wrap(final E a) {
        return null;
    }

    public static <F> Convert<F> create(final F initial) {
        return null;
    }
}
```

javac compiles fine on the snippet above, but ecj reports an error about `Type mismatch: cannot convert from Convert<Object> to Convert<B>` on `res2`.

**Comment by testforstephen** [comment]
@stephan-herrmann Any thoughts on this?

**Comment by stephan-herrmann** [comment]
This is the error:
```
----------
1. ERROR in Convert.java (at line 7)
	Convert<B> res2 = convert(arg -> wrap(create(arg)), a); // error: Type mismatch
	                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Type mismatch: cannot convert from Convert<Object> to Convert<B>
----------
```
Ergo, the outer `convert` invocation seems to pass inference, but with a wrong result.

----
First thing to check: what is the result of `infCtx18.inferInvocationType()` as invoked from `ParameterizedGenericMethodBinding.computeCompatibleMethod18()`.

For the first thing without `wrap()` we get:
```
Type Bounds:
	TypeBound  C#0 :> B
	TypeBound  C#0 = B
	TypeBound  D#1 = B
```
For the variant with `wrap` we actually get a `null` result, i.e., invocation type inference _failed_, `hasReturnProblem` is true, and we provoke the type error by working with the insufficient `provisionalResult` (from invocation applicability inference).

----
Next question: why did that inference fail? 
Again I let the debugger stop right before entering `infCtx18.inferInvocationType()` for the invocation in question (hint: remove the other, successful invocation for simplified debugging :) ). Step into `inferInvocationType()` but disable the 1st breakpoint in order not to be distracted by recursive invocations. 
In a first iteration I see that it fails in the first call `this.currentBounds.incorporate(this)`.
=> Start again and step into `incorporate`

By placing a breakpoint inside `BoundSet.incorporate(InferenceContext18, TypeBound[], TypeBound[])` right inside the block `if (newConstraint != null) { ... }`, I can see that we create a new constraint
```⟨Convert<java.lang.Object> <: Convert<B>⟩```
which is derived from these two:
* `TypeBound  E#3 <: Convert<B>`
* `TypeBound  E#3 :> Convert<java.lang.Object>`

Trying to reduce the new constraint yields `FALSE`, which causes inference to fail immediately.

----
From the above I can guess that `TypeBound  E#3 :> Convert<java.lang.Object>` should be avoided (since it blocks inferring a better type argument to `Convert`). So the next question is: what created this type bound?

Let's teleport the debugger to that location using a breakpoint at the end of TypeBound's constructor, with this condition `toString().equals("TypeBound  E#3 :> Convert<java.lang.Object>")`.
Of course this happens during the nested inference for `wrap(create(arg))` (applicability inference, to be precise). In fact that inference is nested within the outer invocation type inference.

With a bit of luck I could see that the type `Convert<Object>` enters the picture within `ConstraintExpressionFormula.reduce(InferenceContext18)` within the block `if (innerCtx == null) { ... }`. Aha, we never performed any inner inference for `create(arg)`?

Inspecting the current call stack, I can see that we are within `LambdaExpression.cachedResolvedCopy(TypeBinding, boolean, boolean, InferenceContext18, Scope)` and in fact the entire sub-AST we are working on has just been created/copied within this method. 

This implies the problem should be searched directly within `LambdaExpression.resolveType()` when invoked on that fresh AST copy.

Good, next breakpoint within `ReturnStatement.resolve()` right before the call `this.expression.resolveType(scope)`, with condition `this.expression.resolveType(scope)` :)

----

Eventually, we'll enter `PGMB.computeCompatibleMethod18()` for invocationSite `create(arg)`. To avoid the `innerCtx==null` observed above, we need to reach this line:
```
((Invocation) invocationSite).registerInferenceContext(methodSubstitute, infCtx18)
```
We get close, but registering is prevented by `!allArgumentsAreProper` :(

Bingo: by making that finally block unconditional (comment `if (allArgumentsAreProper) {` ) I can make this test pass!

But little surprise: this change would create regressions elsewhere.

----

To sum up today's session: this situation is in conflict with a condition added in commit 51ed0947 on behalf of [Bug 511252](https://bugs.eclipse.org/511252).

**Comment by stephan-herrmann** [comment]
**Meta-comment:** I'm aware that the above could be a bit overwhelming. Still it would be great if you try to replay my debugging session for something like a guided tour through parts of the inference & resolve story. I recommend to use this tour-book during the process: https://wiki.eclipse.org/JDT_Core_Programmer_Guide/ECJ
 :smile:

**Comment by stephan-herrmann** [comment]
I fear that JLS will not give an answer how this case should be handled, given we are in a code area where extra-constitutional stuff is needed in order to be compatible with javac, see https://bugs.eclipse.org/511252#c17

OTOH, that comment was written in 2017. I cannot rule out that JLS has evolved under foot, maybe some change which was not announced to us would actually fix the problem.

**Comment by testforstephen** [comment]
> **Meta-comment:** I'm aware that the above could be a bit overwhelming. 

Not at all, I really like the details of your investigation, thanks for the writeup. I ended up replaying your debugging session step by step and it really helped me sort out how ecj implements the type inference specification.
